### PR TITLE
feat(eth-staking): implement staking logic

### PR DESCRIPTION
**Pull request by StanCheb**
## Description

The staking action logic is implemented. 
The "Stake" transactions list is added to the "Staking" tab.
Fee calculations are corrected.

Tx presented in screenshots: https://goerli1.trezor.io/tx/0xf8c31eb522b200c7d33d628e9b4723072313f572023376dbe947f3c356d0b204

NB! Do not test with real ETH, since unstaking and claiming are not implemented yet. Goerli has only partial support in the wallet SDK. It's recommended to test with Holesky, unless more edge cases with errors are the subject of testing.

## Related Issue

Resolve 
https://github.com/trezor/trezor-suite/issues/9267
https://github.com/trezor/trezor-suite/issues/9257
https://github.com/trezor/trezor-suite/issues/9255
https://github.com/trezor/trezor-suite/issues/9220

## Screenshots:
Знімок екрана 2024-01-11 о 16 06 30
Знімок екрана 2024-01-11 о 16 06 41
Знімок екрана 2024-01-11 о 16 06 49
Знімок екрана 2024-01-11 о 16 07 01
Знімок екрана 2024-01-11 о 16 07 32

### CODE REVIEW

File: suite-common/suite-utils/src/stake.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export const isStakeTx = (txName: string | undefined) => {
+    if (!txName) return false;
+
+    const stakeTxNames = ['Stake', 'Unstake', 'Claim'];
```
Comment (tomasklim) [inline suite-common/suite-utils/src/stake.ts]: It would be better to use `Bytes Signatures` here.

`0x3a29dbae` for staking
`0x33986ffa` for claim
...

`tx.ethereumSpecific?.parsedData?.methodId`

File: packages/suite/src/constants/suite/ethStaking.ts [hunk 2]
```diff
@@ -3,6 +3,8 @@ import BigNumber from 'bignumber.js';
 export const MIN_ETH_AMOUNT_FOR_STAKING = new BigNumber(0.1);
 export const MIN_ETH_FOR_WITHDRAWALS = new BigNumber(0.03);
 export const MIN_ETH_BALANCE_FOR_STAKING = MIN_ETH_AMOUNT_FOR_STAKING.plus(MIN_ETH_FOR_WITHDRAWALS);
+// Source is used for wallet sdk txs to define what app requests a tx.
```
Comment (tomasklim) [inline packages/suite/src/constants/suite/ethStaking.ts]: I do not understand the comment, could you please improve it? I would add link to docs. However, it is not explained there either 

https://wallet-sdk.everstake.one/guide/ethereum/stake.html

File: packages/suite/package.json [hunk 3]
```diff
@@ -19,6 +19,7 @@
         "test-unit:watch": "jest -o --watch"
     },
     "dependencies": {
+        "@everstake/wallet-sdk": "^0.3.37",
```
Comment (tomasklim) [inline packages/suite/package.json:22]: This library should be moved to `@trezor/connect`. 
All the crypto logic (ethereum, cardano, solana,...) should be located there

File: packages/suite/src/utils/suite/stake.ts [hunk 4]
```diff
@@ -30,3 +37,69 @@ export const getStakeFormsDefaultValues = ({
...
+    | {
+          success: true;
+          tx: EthereumTransaction;
+      }
+    | {
+          success: false;
+          errorMessage: string;
+      };
+
+export const prepareStakeEthTx = async ({
```
Comment (tomasklim) [inline packages/suite/src/utils/suite/stake.ts:71]: I would move this part of file to `@trezor/connect` together with the everstake sdk

=> `src/api/ethereum`

File: packages/suite/src/actions/wallet/stake/stakeFormEthereumActions.ts [hunk 5]
```diff
@@ -98,37 +100,35 @@ export const composeTransaction =
...
+                error: 'INCORRECT-FEE-RATE',
+                errorMessage: { id: 'TR_GENERIC_ERROR_TITLE' },
+                type: 'error',
             },
-        });
+        };
+        try {
+            Ethereum.selectNetwork(getEthNetworkForWalletSdk(account.symbol));
+            // TODO: Implement tx type switcher
+            txData = await Ethereum.stake(account.descriptor, amount, WALLET_SDK_SOURCE);
```
Comment (tomasklim) [inline packages/suite/src/actions/wallet/stake/stakeFormEthereumActions.ts:121]: Move to Connect

**Review by tomasklim — CHANGES_REQUESTED** [review]

**Comment by tomasklim** [comment]
Unfortunately, I wasn't able to stake

1. If `10%` is less than the required amount, I would disable the 10% button with some tooltip (applies also to others)
Screenshot 2024-01-15 at 12 10 57

2. If I click on `max` it sets max minus 0.03 in the input but in the review modal, there is max amount and not the max - 0.03
Screenshot 2024-01-15 at 22 00 52

And if I confirm everything on Trezor and start to stake, I get this error
Screenshot 2024-01-15 at 12 24 45

3. When I enter whatever amount except max, I cant get to the next step. `Confirm & stake` does not work for me
Screenshot 2024-01-15 at 21 58 36

4. If I click on `max` and then I change the value, it is not changed in the review modal

https://github.com/everstake/trezor-suite/assets/33235762/7efe803b-53c4-4e31-9c05-69914d284303

5. Staking does not work in desktop app as there is some request to your backend which is not whitelisted
Screenshot 2024-01-15 at 21 59 03

Screenshot 2024-01-15 at 21 58 55

6. Disable or hide the fiat input for testnet as we do not know rates (it is hidden in send form). Util `isTestnet`
Screenshot 2024-01-15 at 12 27 42

7. On Sepolia, I am getting this error (maybe it is not supported)

Screenshot 2024-01-15 at 12 12 18

8. Hide fiat value for testnets, we have util `isTestnet`

Screenshot 2024-01-15 at 22 14 57

9. If I get back to form from the review modal, I do not see clear button
Screenshot 2024-01-15 at 22 19 13

File: suite-common/suite-utils/src/stake.ts (hunk 1 above)
Comment (StanCheb) [reply suite-common/suite-utils/src/stake.ts]: https://github.com/everstake/trezor-suite/pull/7/commits/93e57e6503e1b0ec2120169aa444c7dcc621a7b2

File: packages/suite/src/constants/suite/ethStaking.ts (hunk 2 above)
Comment (StanCheb) [reply packages/suite/src/constants/suite/ethStaking.ts]: https://github.com/everstake/trezor-suite/pull/7/commits/93e57e6503e1b0ec2120169aa444c7dcc621a7b2

**Comment by StanCheb** [comment]
> Unfortunately, I wasn't able to stake
> 
> 1. If `10%` is less than the required amount, I would disable the 10% button with some tooltip (applies also to others)
> 
> Screenshot 2024-01-15 at 12 10 57
> 2. If I click on `max` it sets max minus 0.03 in the input but in the review modal, there is max amount and not the max - 0.03
> 
> Screenshot 2024-01-15 at 22 00 52
> And if I confirm everything on Trezor and start to stake, I get this error Screenshot 2024-01-15 at 12 24 45
> 
> 3. When I enter whatever amount except max, I cant get to the next step. `Confirm & stake` does not work for me
> 
> Screenshot 2024-01-15 at 21 58 36
> 4. If I click on `max` and then I change the value, it is not changed in the review modal
> 
>  Screen.Recording.2024-01-15.at.12.25.56.mov 
> 5. Staking does not work in desktop app as there is some request to your backend which is not whitelisted
> 
> Screenshot 2024-01-15 at 21 59 03 Screenshot 2024-01-15 at 21 58 55
> 6. Disable or hide the fiat input for testnet as we do not know rates (it is hidden in send form). Util `isTestnet`
> 
> Screenshot 2024-01-15 at 12 27 42
> 7. On Sepolia, I am getting this error (maybe it is not supported)
> 
> Screenshot 2024-01-15 at 12 12 18
> 8. Hide fiat value for testnets, we have util `isTestnet`
> 
> Screenshot 2024-01-15 at 22 14 57
> 9. If I get back to form from the review modal, I do not see clear button
> 
> Screenshot 2024-01-15 at 22 19 13

1. Done
2. Fixed
3. Not sure what the problem is. The only case when I had a similar behaviour, was when the emulator was disconnected from the app.
4. Fixed
5. Not sure about this one. I thought we were implementing the web part only. Need some clarification with our backend team
6. This is left for testing purposes. It uses actual ETH rates. Will remove that in the very end, when everything is tested.
7. Sepolia is not supported in the SDK. Removed it from available for staking tokens.
8. This is left for testing purposes. It uses actual ETH rates. Will remove that in the very end, when everything is tested.
9. Not sure how you got this state without getting the field «dirty». You also don’t have the fiat input filled, which is strange. Could you provide more details. please? Added a blind fix for now.

https://github.com/everstake/trezor-suite/pull/7/commits/93e57e6503e1b0ec2120169aa444c7dcc621a7b2

**Comment by tomasklim** [comment]
1. Nice. I would not disable `Max` button and I would add tooltip on hover to disabled buttons `Minimum amount to stake is 0.1 ETH` 

2. ok

3. see code comment below

4. still there, same reason as number 3, `setMaxOutputId` is not reset because of missing `currentRate`

5. There is almost no difference between web and desktop app, don't worry. The domain just needs to be added to whitelist in `allowedDomains` array in `packages/suite-desktop-core/src/config.ts`. It should also work with tor

6. aha oka, probably same reason number 3

7. ok

8. ok

9. fixed

**10. NEW**

https://github.com/everstake/trezor-suite/assets/33235762/cc1c5bd6-dc38-4e8d-9421-80b3fe3417a6

File: packages/suite/src/hooks/wallet/useStakeEthForm.ts [hunk 6]
```diff
@@ -181,9 +198,11 @@ export const useStakeEthForm = ({ selectedAccount }: UseStakeFormsProps): StakeC
         async (amount: string) => {
             setIsAmountForWithdrawalWarningShown(false);
             setIsAdviceForWithdrawalWarningShown(false);
-            if (!fiatRates || !fiatRates.current) return;
+            if (!currentRate) return;
```
Comment (tomasklim) [inline packages/suite/src/hooks/wallet/useStakeEthForm.ts:201]: This condition is a reason for number 3., there is no current rate for testnet coins

Comment (tomasklim) [reply packages/suite/src/hooks/wallet/useStakeEthForm.ts:201]: https://github.com/everstake/trezor-suite/assets/33235762/f1e2ba76-37c7-47d8-a7a0-ce31fa0a5a33

File: packages/suite/package.json (hunk 3 above)
Comment (tomasklim) [reply packages/suite/package.json:22]: Let's do this in second iteration as we discussed

**Comment by StanCheb** [comment]
1. Done.

3 & 4. That’s interesting. Can’t figure out how it’s possible. I use` mapTestnetSymbol() `to get the eth symbol for current rates. Here is this part:

https://github.com/everstake/trezor-suite/pull/7/files#diff-da2957c4b8ba4ab8317bee25a5c27ecefa9d55be5d6da364a287a3489307fdf8R58


https://github.com/everstake/trezor-suite/assets/42133844/ab583c23-475d-4932-b01b-935c7856dc22



Could you, please, check if you have any local changes?

5. Added

10. Another strange case. The message says 0.000003, though it should be 0.03. I wonder how that happened. I didn’t change that constant. The field is supposed to be invalid and empty in this case, since we are unable to calculate max correctly. Is this case reproducible with Goerli or Holesky?

Знімок екрана 2024-01-17 о 16 49 43


https://github.com/everstake/trezor-suite/pull/7/commits/b5fddd3335065bdb802ed3d8a2b431b425eed068

**Comment by StanCheb** [comment]
UPD

I've managed to get a similar behaviour with empty `currentRate` (points 3 & 4). I get it when there are some problems with websockets. But I'm still not quite sure with your case, since you didn't have any errors in the console regarding websockets. Moreover, I'm still able to proceed with staking even without rates

Знімок екрана 2024-01-18 о 14 05 47


https://github.com/everstake/trezor-suite/assets/42133844/12cc6d26-0d6b-445c-9d11-fa49533ac996

**Comment by tomasklim** [comment]
> 1. Done.
> 
> 3 & 4. That’s interesting. Can’t figure out how it’s possible. I use`mapTestnetSymbol()`to get the eth symbol for current rates. Here is this part:
> 
> https://github.com/everstake/trezor-suite/pull/7/files#diff-da2957c4b8ba4ab8317bee25a5c27ecefa9d55be5d6da364a287a3489307fdf8R58
> 
>  2024-01-17.16.06.28.mov 
> Could you, please, check if you have any local changes?
> 
> 5. Added
> 6. Another strange case. The message says 0.000003, though it should be 0.03. I wonder how that happened. I didn’t change that constant. The field is supposed to be invalid and empty in this case, since we are unable to calculate max correctly. Is this case reproducible with Goerli or Holesky?
> 
> Знімок екрана 2024-01-17 о 16 49 43
> [b5fddd3](https://github.com/everstake/trezor-suite/commit/b5fddd3335065bdb802ed3d8a2b431b425eed068)

Sorry, I had local changes of amount, so number 10 is irrelevant, sorry

**Comment by tomasklim** [comment]
> UPD
> 
> I've managed to get a similar behaviour with empty `currentRate` (points 3 & 4). I get it when there are some problems with websockets. But I'm still not quite sure with your case, since you didn't have any errors in the console regarding websockets. Moreover, I'm still able to proceed with staking even without rates
> 
> Знімок екрана 2024-01-18 о 14 05 47
>  2024-01-18.14.20.54.mov

Regarding this, we are having a DDoS attack our ETH backends, so maybe that's the issue with errors. Fiat rates are working for me now 😀

File: packages/suite-desktop-core/src/config.ts [hunk 7]
```diff
@@ -25,6 +25,9 @@ export const allowedDomains = [
     'trezor-cardano-mainnet.blockfrost.io',
     'trezor-cardano-preview.blockfrost.io',
     'blockfrost.dev',
+    'eth-goerli.public.blastapi.io',
+    'ethereum-holesky.publicnode.com',
+    'mainnet.infura.io/v3/f583d4f04d384b9e8c59a7ff1c9f68f1',
```
Comment (tomasklim) [inline packages/suite-desktop-core/src/config.ts]:
```suggestion
    'mainnet.infura.io',
```

Comment (tomasklim) [reply packages/suite-desktop-core/src/config.ts]: _(As we discussed, those backends are temporary. I raised a question if we can provide RPC using our blockbook, if needed)_

**Comment by tomasklim** [comment]
Two last things, see provided video

1. If balance is bigger than `0.1` but the max is not enough (0.13), an error occures
2. if i insert amount which is more than `amount + fee`. It does not show the fee and it allows me to continue. However, continue button in the confirmation modal does nothing
https://github.com/everstake/trezor-suite/assets/33235762/fa845d78-ecef-4465-a723-f1c9923db2b8

As it is kinda edge case, if you want to solve it later, I will approve it after change of infura ⬇️

**Review by tomasklim — CHANGES_REQUESTED** [review]

File: packages/suite-desktop-core/src/config.ts (hunk 7 above)
Comment (StanCheb) [reply packages/suite-desktop-core/src/config.ts]: https://github.com/everstake/trezor-suite/pull/7/commits/6f35bf0e58d70245d04b4ced2b5d2af3d7bf2558

**Comment by StanCheb** [comment]
> Two last things, see provided video
> 
> 1. If balance is bigger than `0.1` but the max is not enough (0.13), an error occures
> 2. if i insert amount which is more than `amount + fee`. It does not show the fee and it allows me to continue. However, continue button in the confirmation modal does nothing
>    https://github.com/everstake/trezor-suite/assets/33235762/fa845d78-ecef-4465-a723-f1c9923db2b8
> 
> As it is kinda edge case, if you want to solve it later, I will approve it after change of infura ⬇️

https://github.com/everstake/trezor-suite/pull/7/commits/6f35bf0e58d70245d04b4ced2b5d2af3d7bf2558

1. Yep, you’re right. I try to compose a tx with the right amount which in the case of clicking max is `balance - 0.03` at this stage. SDK gets the value less than 0.1 and throws an error. As a result we cannot calculate the gas limit and correct max amount, therefore there is nothing to set in the field. That’s why I disabled the max button when` balance - 0.3 < 0.13` before last changes. Returned that logic back.




2.It seems like the field validation is failing to work. In this case you’re supposed to get «Not enough funds» message. 0.104 is too much since `0.104 + fee` will result in total `amount > balance`. There is check against it.

I’ve tested the case with exactly same amount and fee that you have in the video. It seems to work fine for me. Could you, please, check if you have any local changes in `calculate()` function in `stakeFormEthereumActions` (this is where the check with balance happens) or maybe in compose logic on higher levels? Here’s the line: https://github.com/everstake/trezor-suite/pull/7/files#diff-d9ca54e1a6f65e4092fd0072f507d93168ede669cd875195707e2d7fbd5ee4f6L54


https://github.com/everstake/trezor-suite/assets/42133844/e1b801ff-f822-4635-a65d-7519729edc6f



If I emulate your case by returning in the mentioned above `calculate()` some unhandled error, then I can see the case when a user is stuck on the confirm modal. This happens because `composedTx` in `signTx()` is not of type final or undefined. So you can also check what you receive in as `composedTx` in `that signTx()`. Here’s the line: https://github.com/everstake/trezor-suite/blob/feat/stake-eth-action/packages/suite/src/hooks/wallet/useStakeEthForm.ts#L342

**Comment by tomasklim** [comment]
I don't have any local changes.

https://github.com/everstake/trezor-suite/assets/33235762/ace70017-110c-4c27-8adf-9ff33eec4d72

The problem is that `formValues.selectedFee` is changed to `custom` and there is `predefinedLevels` with fee `custom` which is lower than `normal`

Probably added here https://github.com/everstake/trezor-suite/pull/7/files#diff-d9ca54e1a6f65e4092fd0072f507d93168ede669cd875195707e2d7fbd5ee4f6R169


Screenshot 2024-01-20 at 19 26 52


`errors` from `formState` is empty. However, `transactionInfo` from `composedLevels` with `custom` fee has error.
`transactionInfo` is used to show `Maximum fee` in the form. `errors` (`formIsValid`) for `Continue` button
Screenshot 2024-01-20 at 19 35 07

**Comment by StanCheb** [comment]
Is the behaviour the same with Holesky?

**Comment by tomasklim** [comment]
I think the problem is not on Holesky as the gas price is `1`

**Comment by StanCheb** [comment]
I’ve managed to test with real ETH and finally reproduced the problem. So, the problem was indeed in programmatic switching from `normal` to `custom` fee. It tries to switch to the nearest possible fee when there’s an error with `normal` fee. 

This doesn’t happen with devnet because `maxFee` never greater than `minFee` in `composeTransaction()` and thus the `custom` fee object never created. With real ETH it is created and switched to when `normal` fee calculations throw errors. 

I disabled switching to nearest available fee in `useStakeCompose` as well as forming custom fee in `stakeFormEthereumActions`. Left the code commented for future implementations of custom fees.

https://github.com/everstake/trezor-suite/pull/7/commits/ac6e548baa4396c020fab3aaea9a6db672b42315

**Review by tomasklim — APPROVED** [review]
